Gas Attack: Add basic gas price management

Shadowfiend opened the pull request:
The basic mechanism now is:
- We track this transaction's nonce vs the highest pending transaction nonce.
- We also track the number of ticks a transaction has been waiting (each tick is 30s).
- We multiple ticks by the number of transactions ahead of this one, and then apply a 20% factor increase for unit from that multiplication. This has the effect of significantly boosting the gas price every time a new transaction is blocked by the current one.

To handle reboots, error handling has changed so that if we get a `transaction underpriced` or `already known` error, we boost the gas price until we don't see this. That allows a newly booted instance that's resubmitting older transactions to reach the same state as the previous instance. And finally, to handle instances that are catching up to latest chain state, failing to find an LCA after 5 attempts (which are time-based rather than tx-based right now) doesn't crash; instead, it retries after a 60s pause.

Finally, max gas is now 120 gwei.

Shadowfiend committed:
Add super-basic gas management to ethereum/shared.py

Linear increase based on still-pending transaction when a new one is
submitted, and replacing older transactions to the new gas price as
well.

lukasz-zimnoch commented on maintainer/maintainer/ethereum/shared.py:
```diff
@@ -94,7 +95,7 @@ async def sign_and_broadcast(
     logger.info(f'dispatched transaction {tx_id}')
     if not ignore_result:
-        await _track_tx_result(tx_id)
+        asyncio.ensure_future(_track_tx_result(tx, tx_id))
```
I understand why do you want to remove this `await` but I have some doubts. I didn't add it here to guarantee transaction ordering (this is done by the awaits on the upper level) but to make sure we wait until the transaction is actually mined before we trigger subsequent transactions. This way we ensure those subsequent transactions work with a fresh chain state. 

If we remove this `await` I think we may run into the following problem:

1.  We trigger transaction `A` which modifies  some chain state, let's call it `X` 
2. We don't wait for transaction `A` receipt so we trigger transaction `B` whose input consists of chain state `X` Also, suppose transaction `B` calls some code which validates their input on-chain. 
3. Transaction `A` is mined and state `X` is modified
4. Transaction `B` arrives but their input is no longer valid because `X` changed so this transaction is reverted.

IIRC, I observed something similar when the relay invoked `addHeaders` and then `markNewHeaviest` without waiting for the receipt of `addHeaders` transaction. In such case, `markNewHeaviest` reverted because one of their `require` conditions was not satisfied.

Shadowfiend replied:
My reading of the code here is that it doesn't *read* Ethereum state, it only *writes* it. It *reads* Bitcoin state, and writes it to Ethereum. So as long as sequencing of the transactions is preserved, the transactions themselves should be fine. In particular, my reading here was that, before the `await`s, we had no guarantee that `_update_best_digest` would dispatch its transaction with a higher nonce than any preceding `_add_headers` or `_add_diff_change` transactions---it depended on how `asyncio` scheduled those calls. So we could `_update_best_digest` on Ethereum without having up-to-date header/diff-change info, and this would fail.

If that reading is true, then order-of-nonce is all that matters, and waiting for executions in order does not. Does it seem true?

lukasz-zimnoch replied:
Mostly true but the specific part I'm worried about is `_update_best_digest`. Here we prepare the `markNewHeaviest` function's input by calling `contract.get_best_block()` which corresponds to `getBestKnownDigest` chain function and then `find_lca` which calls the `isAncestor` function from `Relay` contract. If you look closer to `isAncestor` you can see it uses `previousBlock` which is in turn modified by `addHeaders` function. There is a high chance this is no longer a problem since we ensure correct nonce order but I just want to draw your attention here and make sure we don't return to the same problems we solved recently.

Shadowfiend replied:
Perfect, that's the piece I missed! I'll give it a deeper look and make sure this behaves as expected, or adjust if not.

Shadowfiend replied:
So I think this would generally be guarded by the [preflight](https://github.com/keep-network/relays/blob/bc881b9b503cbdbfdfd8ff3f398fcbc18132f6d9/maintainer/maintainer/header_forwarder/push.py#L177-L186). The caveat is that if we're moving through older blocks quickly (because we're catching up from our initial parameters, for example), that might not work 100% of the time. In those cases, though, I expect the restart would be sufficient, as it should result in fresher state.

I'm thinking we can make this even safer by changing `'latest'` to `'pending'` [in the `get_best_block` call](https://github.com/keep-network/relays/blob/bc881b9b503cbdbfdfd8ff3f398fcbc18132f6d9/maintainer/maintainer/ethereum/contract.py#L94). Thoughts?

lukasz-zimnoch replied:
Yeah, sounds good!

Shadowfiend replied:
Upon further investigation, in cases where the relay is submitting transactions with insufficient gas price, we end up stacking pending transactions. These transactions have consistently cleared without error eventually, so it doesn't seem we need to change to `pending` in this lookup.

sthompson22 commented on maintainer/maintainer/ethereum/shared.py:
```diff
@@ -159,16 +167,42 @@ def _adjust_gas_price(gas_price: int) -> int:
+                    data = tx.data,
+                    chainId = tx.chainId)
+
+                # Bradcast and set up tracking for the new tx, and stop watching
```
Bradcast -> Broadcast

Shadowfiend replied:
Listen, Brad can do whatever he wants.

Shadowfiend committed:
Start INCOMPLETE_TX_COUNT at pending transaction count

This assumes the account operating the relay is not doing anything else.

Shadowfiend committed:
Replace pending relay transactions on boot

When the relay restarts, it's willing to replace pending transactions
with a higher gas price replacement in order to speed up catch-up times.

Shadowfiend committed:
Track completed/pending nonce to use for gas boost

The adjustment is now based on the difference between any given
transaction's nonce and the next one up. This helps transactions that
have been waiting for longer get a boosted price, while more recent
transactions wait longer if needed. It also ensures the price drops back
down to the default price when the default price becomes reasonable
again.

Shadowfiend committed:
Properly compute initial gas price

When there are other pending transactions, the gas price was always
being set to 1 initially. This fixes that, setting it to a high value if
there are many pending transactions ahead.

Shadowfiend committed:
Add some nonce/pending transaction info logging

Shadowfiend committed:
n is not a nonce

But it is a variable we used to know.

Shadowfiend committed:
Fix an off-by-one for starting nonce

Shadowfiend committed:
await sign_and_broadcast in _track_tx_result

Strictly speaking it doesn't make a huge difference, but it handles a
Python warning.

Shadowfiend committed:
Don't try catching up nonces if already caught up

Shadowfiend committed:
Boost gas price even if a transaction waits alone

Ideally the relay doesn't pile up unconfirmed transactions at all. Count
ticks so that, as a transaction sits for longer, it is replaced by
gradually increased gas costs.

Also update default gas price to 20, which is more reasonable on
mainnet, and adjust the curve to increase by 20% increments instead of
by full DEFAULT_GAS_PRICE increments.

Shadowfiend commented:
Ok, so the basic updates here are:
- We've shifted to tracking current transaction nonce vs latest pending transaction nonce.
- We also track the number of ticks a transaction has been waiting (each tick is 30s).
- We sum these, and then apply a 20% factor increase for each tick or pending transaction that is backed up in front of the current transaction.
- Every tick, if the computed price based on this info is higher than the one that's currently being tracked for the transaction, we resubmit it with a boosted price.
- We also consider the number of pending transactions when we submit a new one, and use that to adjust the gas price if needed. This is useful when the maintainer reboots, to boost older transactions that might still be in the mempool immediately.

Shadowfiend committed:
Drop LATEST_COMPLETED_NONCE global

It's no longer needed, as the transaction nonce of the transaction under
consideration is the one used to make decisions about gas price, rather
than the overall latest completed transaction.

Shadowfiend committed:
Ensure adjusted_gas_price is an integer

Shadowfiend committed:
Fix a comment typo, Brad does not cast

Shadowfiend committed:
Fix a bad reference to UnsignedTx.gasPrice

Shadowfiend committed:
Close the loop for ongong gas price increases

tick information is passed along with each new transaction submission,
and tx results are tracked all the way down. Finally, "known
transaction" errors are caught and retried with a higher gas price after
a wait.

Still missing: handling _underpriced_ transactions.

Shadowfiend committed:
Add additional nonce/gas price logging

In cases where a transaction fails to go through, some logging clarifies
the nonce and gas price information.

Shadowfiend committed:
Underpriced transactions go down the gas-bump path

Since underpriced transactions indicate the submitted transaction for
that nonce has too low a gas price, entering the gas-bump loop allows
the gas bumping to get the transaction to an appropriate price.

To do this, _track_tx_result can now operate without a tx_id. This is
because, in the case of underpriced transactions, no tx_id is available
to pass along. A lack of tx_id is treated as a transaction needing
retry.

Shadowfiend committed:
Track gas-bumped transactions

Resubmissions with higher gas prices were not being tracked, so only one
gas bump could occur. These are now properly tracked for success, so
that they can have their gas price bumped if they don't succeed in a
reasonable amount of time.

Shadowfiend committed:
Don't crash if an LCA isn't found in 5 attempts

Instead, wait 60 seconds and then try again. This is normal when gas
prices require multiple bumps to clear the related transactions.

Shadowfiend committed:
Log mined tx count at startup

Shadowfiend committed:
Aggressively clear unmined txes during gas bumping

Before, there was an equivalent gas price bump for remaining unmined for
30 seconds and for having additional transactions stuck behind it. Now,
having additional transactions stuck behind a transaction gives it a
much higher gas bump (2^n instead of just n), ensuring that transactions
that are blocking the contract clear quickly.

Shadowfiend committed:
Fix mined_tx_count

This number was actually mined_tx_count + 1, apparently. This also lets
us drop an if statement, simplifying code flow.

Shadowfiend committed:
Handle already known errors like transaction underpriced errors

Additionally, this scenario is logged at WARN level so there's a trace
that the maintainer is boosting gas in this case.

Shadowfiend committed:
Do not crash on nonce too low errors

This typically happens because a lower-priced version of the same
transaction cleared while a higher-priced version was about to be
submitted. Rather than bomb out, the transaction submission is simply
skipped in that case.

Shadowfiend committed:
Include nonce + gas price info in a few logs

This makes for easier correlation of logs during as boosts when multiple
transactions are in the process of retrying.

Shadowfiend committed:
Gas price factor is now (txes ahead) * ticks

This makes a new transaction blocked ahead of this one boost the impact
of the number of ticks this transaction has been waiting.

Shadowfiend committed:
When catching up to existing txes, don't wait 30s

Before, boosting gas would wait 30s at each tick, even if there was
already a tx at this nonce and this or higher gas price. This wait is
now gone, so if there's a catch-up to be done it happens in a tight
loop.

Shadowfiend committed:
Switch an await to an ensure_future

This shouldn't have much of an effect in practice, but is a clearer
expression of the intent here.

Shadowfiend committed:
Add logging and retries when boosting gas to match existing tx

When the maintainer restarts, its transaction resubmissions start at 1
tick and have to catch up to the number of ticks of the existing pending
transactions on chain, if any. In these cases, the new gas price can
end up below the latest gas price. For these situations, the gas price
is now boosted until it either reaches MAX_GAS_PRICE or exceeds the
latest gas price.

Additionally, irrespective of how the gas bump path is handled, the path
through it is logged.

Shadowfiend committed:
Set max gas to 120 gwei

Unfortunately, this is a necessity in the current gas price climate.

Shadowfiend commented:
@lukasz-zimnoch this is ready for another look.

lukasz-zimnoch reviewed (approved):
LGTM! If @sthompson22 and @pdyraga don't have additional comments I think we can merge.
